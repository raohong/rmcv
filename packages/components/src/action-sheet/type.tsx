export type ActionSheetAction = {
  /**
   * @description 标题
   */
  name: React.ReactNode;
  /**
   * @description 二级标题
   */
  subName?: React.ReactNode;
  /**
   * @description 文字颜色
   */
  color?: string;
  /**
   * @description 选项自定义 class
   */
  className?: string;
  /**
   * @description 是否为加载状态
   */
  loading?: boolean;
  /**
   * @description 是否为禁用状态
   */
  disabled?: boolean;
  /**
   * @description 点击时触发的回调函数
   */
  callback?: (action: ActionSheetAction) => void;
};

export type ActionSheetProps = {
  /**
   * @description 是否显示动作面板
   */
  visible?: boolean;
  /**
   * @description 面板选项列表
   */
  actions?: ActionSheetAction[];
  /**
   * @description 动作面板标题
   */
  title?: React.ReactNode;
  /**
   * @description 动作面板描述
   */
  description?: React.ReactNode;
  /**
   * @description 关闭按钮文字
   */
  cancelText?: React.ReactNode;
  /**
   * @description 是否可关闭的
   * @default true
   */
  closable?: boolean;
  /**
   * @description 关闭图标
   */
  closeIcon?: React.ReactNode;
  /**
   * @description 是否显示圆角
   * @default true
   */
  round?: boolean;
  /**
   * @description 自定义 class
   */
  className?: string;
  /**
   * @description 是否显示 overlay
   * @default true
   */
  overlay?: boolean;
  /**
   * @description overlay 自定义 className
   */
  overlayClassName?: string;
  /**
   * @description overlay 自定义 style
   */
  overlayStyle?: React.CSSProperties;
  /**
   * @description overlay 点击是否关闭
   * @default true
   */
  overlayClosable?: boolean;
  /**
   * @description 是否在显示弹层时才渲染节点
   * @default true
   */
  lazyRender?: boolean;
  /**
   * @description
   * @default true
   */
  lockScroll?: boolean;
  /**
   * @description 是否在页面回退时自动关闭
   */
  closeOnPopState?: boolean;
  /**
   * @description 是否开启底部安全区适配
   */
  safeArea?: boolean;
  /**
   * @description 点击选项后是否关闭
   */
  closeOnClickAction?: boolean;
  /**
   * @description 点击选项时触发，禁用或加载状态下不会触发
   */
  onSelect?: (action: ActionSheetAction, index: number) => void;
  /**
   * @description 关闭回调
   */
  onClose?: () => void;
  /**
   * @description 点击取消按钮，此时还要触发 onClose
   */
  onCancel?: () => void;
  /**
   * @description 点击 overlay 回调
   */
  onOverlayClick?: () => void;
  /**
   * @description 关闭前的回调函数，返回 false 可阻止关闭，支持返回 Promise
   */
  onBeforeClose?: (
    action: ActionSheetAction,
  ) => void | boolean | Promise<boolean>;
  /**
   * @description 动画结束后回调
   */
  afterVisibleChange?: (visible: boolean) => void;
  /**
   * @description 自定义内容
   */
  children?: React.ReactNode;
};
