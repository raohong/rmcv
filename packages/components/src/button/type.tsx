import type { JSXIntrinsicElementProps } from '../types';
import type { LoadingProps } from '../loading';

export type ButtonSize = 'large' | 'default' | 'small' | 'mini';

export type ButtonType = 'primary' | 'success' | 'warning' | 'danger' | 'link';

export type ButtonShape = 'round' | 'square';

export type NativeButtonHTMLType = 'button' | 'submit' | 'reset';

type BaseButtonProps = {
  /**
   * @description 按钮类型
   */
  type?: ButtonType;
  /**
   * @description 按钮尺寸
   */
  size?: ButtonSize;
  /**
   * @description 是否是块级元素
   */
  block?: boolean;
  /**
   * @description 是否镂空
   */
  plain?: boolean;
  /**
   * @description 按钮形状
   */
  shape?: ButtonShape;
  /**
   * @description 按钮 disabled
   */
  disabled?: boolean;
  /**
   * @description 按钮 icon
   */
  icon?: React.ReactElement;
  /**
   * @description 是否 loading
   */
  loading?: boolean;
  /**
   * @description border 是否是高清
   */
  hairline?: boolean;
  /**
   * @@description 点击事件
   */
  onClick?: React.MouseEventHandler<HTMLElement>;
  /**
   * @@description 按钮内容
   */
  children?: React.ReactNode;
  /**
   * @@description loading 图标文字
   */
  loadingText?: string;
  /**
   * @@description loading 图标类型
   */
  loadingType?: LoadingProps['type'];
  /**
   * @@description loading 图标大小
   * @default 20px
   */
  loadingSize?: LoadingProps['size'];
};

type AnchorButtonProps = {
  /**
   * @description 按钮 html 链接
   */
  href?: string;
  target?: string;
} & BaseButtonProps &
  JSXIntrinsicElementProps<'button', 'onClick' | 'type'>;

type NativeButtonProps = {
  /**
   * 按钮的原始 type
   */
  htmlType?: NativeButtonHTMLType;
} & BaseButtonProps &
  JSXIntrinsicElementProps<'button', 'onClick' | 'type'>;

export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>;
