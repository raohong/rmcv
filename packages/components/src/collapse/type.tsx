import type React from 'react';
import type { CellProps } from '../cell';

export type CollapseKey = string | string[];

export type CollapseItemProps = {
  /**
   * @description 对应的 activeKey
   */
  key?: string;
  itemKey?: string;
  toggle?: (itemKey: string | undefined) => void;
  /**
   * @description 是否折叠
   */
  collapsed?: boolean;
  /**
   * @description 内容
   */
  children?: React.ReactNode;
  /**
   * @description 禁用
   */
  disabled?: boolean;
} & Pick<
  CellProps,
  | 'icon'
  | 'size'
  | 'title'
  | 'value'
  | 'label'
  | 'border'
  | 'isLink'
  | 'titleClassName'
  | 'valueClassName'
  | 'labelClassName'
> &
  React.HTMLAttributes<HTMLDivElement>;

export type CollapseProps = {
  /**
   * @description 当前展开
   */
  activeKey?: CollapseKey;
  /**
   * @description 默认展开的 key
   */
  defaultActiveKey?: CollapseKey;
  /**
   * @description 切换面板时触发
   */
  onChange?: (current: string[]) => void;
  /**
   * @description 禁用
   */
  disabled?: boolean;
  /**
   * @description 折叠内容
   */
  children?: React.ReactNode;
  /**
   * @description 手风琴模式
   */
  accordion?: boolean;
  /**
   * @description 边框
   */
  border?: boolean;
} & Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'>;
