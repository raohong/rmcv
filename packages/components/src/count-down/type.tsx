import type React from 'react';
import type { JSXIntrinsicElementProps } from '../types';

export type CountDownTimeData = {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  milliseconds: number;
};

export type CountDownProps = {
  /**
   * @description 倒计结束完成回调
   */
  onFinish?: () => void;
  /**
   * @description 每次计时
   */
  onChange?: (data: CountDownTimeData) => void;
  /**
   * @description 用于自定义内容
   */
  children?: (data: CountDownTimeData) => React.ReactNode;
  /**
   * @description 倒计时时间 以秒为单位
   */
  time?: number;
  /**
   * @description 时间格式
   * DD	天数
   * HH	小时
   * mm	分钟
   * ss	秒数
   * S	毫秒（1 位）
   * SS	毫秒（2 位）
   * SSS	毫秒（3 位）
   * @example DD:HH:mm
   *
   */
  format?: string;
  /**
   * @description 自动开始倒计时
   */
  autoStart?: boolean;
  /**
   * @description 是否开启毫秒级别计时
   */
  millisecond?: boolean;
} & JSXIntrinsicElementProps<'div', 'onChange'>;

export type CountDownRef = {
  pause: () => void;
  start: () => void;
  reset: () => void;
};
