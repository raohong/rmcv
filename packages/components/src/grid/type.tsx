import type { IntrinsicElementsKeys, JSXIntrinsicElementProps } from '../types';

export type GridDirection = 'vertical' | 'horizontal';

export type GridProps = {
  /**
   * @description 列数
   * @default 3
   */
  column?: number;
  /**
   * @description 图标大小
   * @default 28
   */
  iconSize?: number | string;
  /**
   * @description 格子之间的间距
   */
  gutter?: number | string;
  /**
   * @description 是否显示边框
   */
  border?: boolean;
  /**
   * @description 是否将格子内容居中显示
   */
  center?: boolean;
  /**
   * @description 是否将格子固定为正方形
   */
  square?: boolean;
  /**
   * @description 是否开启格子点击反馈
   */
  clickable?: boolean;
  /**
   * @description 格子内容排列的方向
   * @default vertical
   */
  direction?: GridDirection;
  /**
   * @description 渲染组件
   * @default div
   */
  component?: IntrinsicElementsKeys;
} & JSXIntrinsicElementProps<'div'>;

export type GridItemProps = {
  /**
   * @description 文字内容
   */
  text?: React.ReactNode;
  /**
   * @description 图标
   */
  icon?: React.ReactNode;
  /**
   * @description 图标大小
   * @default 28
   */
  iconSize?: number | string;
  /**
   * @description 是否显示徽标小红点
   */
  dot?: boolean;
  /**
   * @description 右上角徽标的内容
   */
  badge?: number | string;
  /**
   * @description 徽标内容为数字时，最大值
   * @default 99
   */
  max?: number;
  /**
   * @description 当 content 是 0 时，是否显示
   * @default true
   */
  showZero?: boolean;
  /**
   * @description 渲染组件
   * @default div
   */
  component?: IntrinsicElementsKeys;
  /**
   * @description content 自定义
   */
  contentClassName?: string;
} & JSXIntrinsicElementProps<'div'>;
