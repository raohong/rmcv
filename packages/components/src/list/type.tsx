import React from 'react';
import type { JSXIntrinsicElementProps } from '../types';
import type { ListLoadingStatus } from './constants';

export type ListRef = {
  check: () => void;
};

export type ListProps = {
  /**
   * @description 加载状态改变时的回调
   */
  onLoadingStatusChange?: (status: ListLoadingStatus) => void;
  /**
   * @description 加载状态
   * @default none
   */
  loadingStatus?: ListLoadingStatus;
  /**
   * @description 滚动条与底部或者顶部距离小于 offset 时触发 load 事
   */
  offset?: number;
  /**
   * @description 加载时为文字
   */
  loadingText?: string;
  /**
   * @description 加载完成后的状态
   */
  finishedText?: string;
  /**
   * @description 加载失败时的文字
   */
  errorText?: string;
  /**
   * @description 是否在初始化时立即执行滚动位置检查
   */
  immediateCheck?: boolean;
  /**
   * @description 加载回调
   */
  onLoad?: () => void | Promise<boolean>;
  /**
   * @description 自定义渲染 error
   */
  renderErrror?: () => React.ReactNode;
  /**
   * @description 自定义渲染 loading
   */
  renderLoading?: () => React.ReactNode;
  /**
   * @description 自定义渲染 finished
   */
  renderFinished?: () => React.ReactNode;
  /**
   * @description 加载完成后是否还能再加载
   * @default true
   */
  disableOnFinished?: boolean;
  /**
   * @description 是否自动设置状态根据 onLoad 触发状态
   */
  autoSetStatusOnLoad?: boolean;
} & JSXIntrinsicElementProps<'div', 'onLoad'>;
