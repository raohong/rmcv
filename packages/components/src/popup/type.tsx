import type { SpringConfig } from '@react-spring/web';

export type PopupPositon = 'left' | 'right' | 'top' | 'bottom' | 'center';
export type PopupCloseIconPosition =
  | 'top-left'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-right';

export type PopupTransitionConfig<T extends object = object> = {
  from: T;
  enter: T;
  leave: T;
  config?: SpringConfig;
};

export type PopupProps = {
  /**
   * @description popup visible
   */
  visible?: boolean;
  /**
   * @description
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * @description popup 位置
   * @default center
   */
  position?: PopupPositon;
  /**
   * @description 是否圆角
   */
  round?: boolean;
  /**
   * @description 是否懒渲染
   */
  lazyRender?: boolean;
  /**
   * @description 自定义 style
   */
  style?: React.CSSProperties;
  /**
   * @description 自定义 class
   */
  className?: string;
  /**
   * @description 是否可关闭
   */
  closeable?: boolean;
  /**
   * @description 关闭按钮
   */
  closeIcon?: React.ReactElement;
  /**
   * @description 关闭按钮位置
   * @default top-right
   */
  closeIconPosition?: PopupCloseIconPosition;
  /**
   * @description 是否显示 overlay
   */
  overlay?: boolean;
  /**
   * @description overlay 自定义 className
   */
  overlayClassName?: string;
  /**
   * @description overlay 自定义 style
   */
  overlayStyle?: React.CSSProperties;
  /**
   * @description
   */
  overlayClosable?: boolean;
  /**
   * @description 自定义渲染位置
   */
  getContainer?: () => HTMLElement;
  /**
   * @description 是否启用 safe-area
   */
  safeArea?: boolean;

  transiton?: PopupTransitionConfig;
};
