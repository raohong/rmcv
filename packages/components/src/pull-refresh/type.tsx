import type React from 'react';
import type { SpringValue, Interpolation } from '@react-spring/core';

export type PullRefreshRenderParams = {
  pullDistance: number;
  headerHeight: number;
  progress: Interpolation<number, number>;
  value: SpringValue<number>;
};

export type PullRefreshProps = {
  /**
   * @description 顶部内容高度
   * @default 50
   */
  headHeight?: string | number;
  /**
   * @description 刷新距离
   * @default 与顶部内容一样
   */
  pullDistance?: string | number;
  /**
   * @description 下拉时内容
   */
  pullingText?: React.ReactNode;
  /**
   * @description 刷新时内容
   */
  loadingText?: React.ReactNode;
  /**
   * @description 释放到刷新之间内容
   */
  loosingText?: React.ReactNode;
  /**
   * @description 刷新后内容
   */
  successText?: React.ReactNode;
  /**
   * @description 下拉成功内容
   */
  successDuration?: number;
  /**
   * @description 渲染非下拉状态时顶部内容
   */
  renderNormal?: () => React.ReactNode;
  /**
   * @description 渲染下拉过程中顶部内容
   */
  renderPulling?: (params: PullRefreshRenderParams) => React.ReactNode;
  /**
   * @description 渲染释放过程中顶部内容
   */
  renderLoosing?: (params: PullRefreshRenderParams) => React.ReactNode;
  /**
   * @description 渲染加载过程中顶部内容
   */
  renderLoading?: (params: PullRefreshRenderParams) => React.ReactNode;
  /**
   * @description 渲染刷新成功提示内容
   */
  renderSuccess?: () => React.ReactNode;
  /**
   * @description 是否禁止
   */
  disabled?: boolean;
  /**
   * @description 触发刷新回调
   */
  onRefresh?: () => Promise<any>;
  /**
   * @description 内容 class
   */
  contentClassName?: string;
} & JSX.IntrinsicElements['div'];

export type PullRefreshRef = {
  refresh: () => void;
};
