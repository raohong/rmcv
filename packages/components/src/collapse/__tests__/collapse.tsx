import { useControllableValue } from '@rmc-vant/hooks';
import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import type { CollapseProps } from '..';
import { Collapse, CollapseItem } from '..';

const testId = 'collapse';

const App: React.FC<Pick<CollapseProps, 'activeKey' | 'onChange'>> = (props) => {
  const [activeKey, setActiveKey] = useControllableValue(props, {
    valuePropName: 'activeKey',
  });
  return (
    <Collapse data-testid={testId} onChange={setActiveKey} activeKey={activeKey}>
      <CollapseItem title='1' data-testid='1' key='1' />
      <CollapseItem title='2' data-testid='2' key='2' />
    </Collapse>
  );
};

it('render correctly', () => {
  const tree = render(<Collapse />);

  expect(tree.asFragment()).toMatchSnapshot();
});

it('render with defaultActiveKey', () => {
  render(
    <Collapse data-testid={testId} defaultActiveKey='1'>
      <CollapseItem data-testid='1' key='1' />
      <CollapseItem data-testid='2' key='2' />
    </Collapse>,
  );

  expect(screen.getByTestId('1')).toHaveAttribute('aria-expanded', 'true');
  expect(screen.getByTestId('2')).toHaveAttribute('aria-expanded', 'false');
});

it('render with activeKey and onChange', async () => {
  const handler = jest.fn();

  const com = render(<App onChange={handler} />);

  expect(screen.getByTestId('1')).toHaveAttribute('aria-expanded', 'false');
  expect(screen.getByTestId('2')).toHaveAttribute('aria-expanded', 'false');

  act(() => {
    fireEvent.click(screen.getByTestId('1').firstElementChild as HTMLElement);
  });

  expect(screen.getByTestId('1')).toHaveAttribute('aria-expanded', 'true');
  expect(handler).toHaveBeenCalledWith(['1']);

  com.rerender(<App onChange={handler} activeKey={[]} />);

  expect(screen.getByTestId('1')).toHaveAttribute('aria-expanded', 'false');

  com.rerender(<App onChange={handler} activeKey={['2']} />);

  expect(screen.getByTestId('1')).toHaveAttribute('aria-expanded', 'false');
  expect(screen.getByTestId('2')).toHaveAttribute('aria-expanded', 'true');
});
