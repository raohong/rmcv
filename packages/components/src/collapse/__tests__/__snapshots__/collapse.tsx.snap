// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
}

<div
    class="RVuiCollapse-root emotion-0"
  />
</DocumentFragment>
`;
